extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaIndexTask.java

```diff
@@ -266,39 +306,24 @@ public void run()
         }
       };
 
-      // Initialize consumer assignment.
-      final Set<Integer> assignment = Sets.newHashSet();
-      for (Map.Entry<Integer, Long> entry : nextOffsets.entrySet()) {
-        final long endOffset = ioConfig.getEndPartitions().getPartitionOffsetMap().get(entry.getKey());
-        if (entry.getValue() < endOffset) {
-          assignment.add(entry.getKey());
-        } else if (entry.getValue() == endOffset) {
-          log.info("Finished reading partition[%d].", entry.getKey());
-        } else {
-          throw new ISE(
-              "WTF?! Cannot start from offset[%,d] > endOffset[%,d]",
-              entry.getValue(),
-              endOffset
-          );
-        }
-      }
-
-      assignPartitions(consumer, topic, assignment);
-
-      // Seek to starting offsets.
-      for (final int partition : assignment) {
-        final long offset = nextOffsets.get(partition);
-        log.info("Seeking partition[%d] to offset[%,d].", partition, offset);
-        consumer.seek(new TopicPartition(topic, partition), offset);
-      }
+      Set<Integer> assignment = assignPartitionsAndSeekToStart(consumer, topic);
 
       // Main loop.
-      // Could eventually support early termination (triggered by a supervisor)
       // Could eventually support leader/follower mode (for keeping replicas more in sync)
       boolean stillReading = true;
       while (stillReading) {
+        if (possiblyPause(assignment)) {
+          // The partition assignments may have changed while paused by a call to setEndOffsets() so reassign
+          // partitions upon resuming. This is safe even if the end offsets have not been modified.
+          assignment = assignPartitionsAndSeekToStart(consumer, topic);
```

Comment:
Is there a need to seek here? The nextOffsets shouldn't change, so I think this could just be an assign rather than assign-and-seek